### ISSUE TITLE
Rewrite in Rust

### ISSUE THREAD

**Issue by sherlockholmestech** [DET171/hd2_app#1]
The decision to transition the HD2_App from Flutter to Rust stems from a myriad of considerations, chief among them being the need for enhanced performance, reliability, and maintainability. While Flutter has proven proficient in delivering flexible and rapid development experiences, there are emerging requirements and considerations that necessitate exploring alternative frameworks. Rust, with its low-level control, efficient memory management, and native code compilation capabilities, emerges as a compelling alternative with the potential to address these challenges and unlock new possibilities for the HD2_App.

One area where Rust excels is in handling asynchronous tasks efficiently. In modern applications, asynchronous programming is crucial for responsiveness and scalability, especially when dealing with network requests or concurrent operations. For example, consider the task of fetching data from a network API asynchronously:

In Flutter, one might use Dart's `async` and `await` keywords with the `http` package:

```dart
Future<void> fetchData() async {
    var response = await http.get('https://api.example.com/data');
    print('Response: ${response.body}');
}
```

In Rust, the same task can be accomplished using `async fn` with `async-std`:

```rust
async fn fetch_data() -> Result<(), reqwest::Error> {
    let response = reqwest::get("https://api.example.com/data").await?;
    let body = response.text().await?;
    println!("Response: {}", body);
    Ok(())
}
```

This demonstrates how Rust's asynchronous capabilities, coupled with its powerful asynchronous runtime libraries, can improve the HD2_App's responsiveness and resource utilization.

Furthermore, Rust's strong emphasis on memory safety and minimal runtime overhead offers distinct advantages, particularly in resource-constrained environments or safety-critical applications. Error handling, a crucial aspect of writing reliable software, is ingrained in Rust's core philosophy, allowing developers to catch and handle errors with precision and confidence. Contrast the error handling approach in Dart, where exceptions are thrown and caught using `try` and `catch` blocks, with Rust's explicit error handling using the `Result` type and pattern matching. This difference not only promotes safer and more predictable code but also reduces the risk of unexpected runtime errors and crashes, enhancing the overall stability and robustness of the HD2_App.

Moreover, Rust's support for concurrency and parallelism enables developers to harness the full power of modern multi-core processors and distributed systems. By leveraging tools like `tokio` or `rayon`, developers can write highly concurrent and parallel code with ease, improving the application's throughput and scalability. This is particularly relevant for the HD2_App, which may need to handle multiple tasks concurrently, such as processing user inputs, fetching data from multiple sources, or performing background computations. Rust's concurrency primitives, such as lightweight threads (or "green threads") and asynchronous tasks, provide a flexible and efficient way to tackle these challenges, ensuring that the HD2_App remains responsive and performant under heavy workloads.

Another area where Rust shines is in file I/O operations, where its asynchronous capabilities can significantly improve performance and responsiveness. While Flutter provides asynchronous file I/O APIs through Dart's `File` class and `dart:io` library, Rust's asynchronous file I/O capabilities, powered by libraries like `async-std` or `tokio`, offer finer-grained control and better integration with Rust's overall asynchronous programming model. This enables developers to perform file operations asynchronously without blocking the application's main event loop, ensuring smooth and responsive behavior even when dealing with large files or slow I/O devices.

Moreover, Rust's prowess extends to GUI development, thanks to libraries like `gtk-rs` and `druid`. While Flutter provides a rich set of UI widgets out of the box, Rust's ecosystem for GUI development, albeit in its nascent stages, is rapidly evolving and maturing. With libraries like `gtk-rs`, developers can create native desktop applications with rich and responsive user interfaces, leveraging the full power of Rust's safety and performance guarantees. This opens up new possibilities for the HD2_App, allowing it to target a wider range of platforms and devices while maintaining a consistent and native user experience.

In conclusion, while the transition from Flutter to Rust entails learning curves and adjustments, the benefits in terms of performance, reliability, and maintainability are profound. Rust's feature-rich ecosystem and robust language design empower developers to tackle a wide array of challenges, making it a compelling choice for modern application development, including the HD2_App project. By leveraging Rust's unique features and capabilities, the HD2_App can not only address current challenges but also future-proof itself against evolving technological trends and user expectations, ensuring its continued success and relevance in the ever-changing landscape of mobile and embedded development.

**Comment by DET171** [comment]
PRs are once again welcome
